### PR TITLE
Add expression coverage on std_logic logic ops.

### PR DESCRIPTION
**Pull request by Blebowski**
Add expression coverage on logic operands between inputs of type std_logic and std_logic_vector.

For std_logic_vector, track each bit of the input operands. Emit only single coverage
OP for whole array, and comparison logic to check for LHS/RHS combinations only once
to avoid bloating of generated VCODE. When f-call for the operation is executed,
do a run-time loop which checks value of each bit.

I use `ignore-arrays-from` to limit length of std_logic_vector operand which is
considered for collection of expression coverage. Possibly, this might be split to
a separate option, but I think it makes sense to keep single option.

Another thing which might be added, is an option to enable / disable collection over
std_logic, but I think its enough to always collect it.

This is more-less last code coverage feature which adds collection of new coverage data
(still FSM coverage might be usefull, but I will not even start with that now, since detection
  of FSMs and reachable transitions between states is non-trivial...).

MR is now ready for review.

### CODE REVIEW

File: src/rt/cover.c [hunk 1]
```diff
@@ -175,6 +175,37 @@ static bool cover_is_branch(tree_t branch)
...
+   struct {
+      ident_t op;
+      unsigned flags;
+   } std_log_ops[] = {
+      { ident_new("\"and\"")  ,COVER_FLAGS_AND_EXPR},
+      { ident_new("\"nand\"") ,COVER_FLAGS_AND_EXPR},
+      { ident_new("\"or\"")   ,COVER_FLAGS_OR_EXPR},
+      { ident_new("\"nor\"")  ,COVER_FLAGS_OR_EXPR},
+      { ident_new("\"xor\"")  ,COVER_FLAGS_XOR_EXPR},
+      { ident_new("\"xnor\"") ,COVER_FLAGS_XOR_EXPR}
```
Comment (nickg) [inline src/rt/cover.c:198]: You can add these to `intern_strings()` in `common.c` and then make the first line of the table a `well_known_t`.

**Review by nickg — COMMENTED** [review]
Rebasing on the latest master should fix the tests.

File: src/lower.c [hunk 2]
```diff
@@ -1742,6 +1772,148 @@ static vcode_reg_t lower_logical(tree_t fcall, vcode_reg_t result,
...
+      return;
+
+   flags |= COV_FLAG_EXPR_STD_LOGIC;
+
+   assert(tree_params(fcall) == 2);
+
+   type_t arg_type = type_param(tree_type(decl), 0);
+
+   if (type_is_array(arg_type)) {
+      assert(type_is_array(type_param(tree_type(decl), 1)));
```
[the remark this answers is not in the document]
Comment (Blebowski) [reply src/lower.c:1832]: Good, point, thanks. I have pushed first implementation with `std_logic_vector` handled.
Two issues are still not fixed there. The one you mentioned, and querying dimensions
from results of another functions.

**Comment by Blebowski** [comment]
I have fixed the two issues you have pointed out.

I tried to simplify the code little bit. In the end, I don't skip the coverage on mismatching lengths
of arrays, but emit least common length, it seemed simpler than checking the matching lengths
when taking into account the vhdl2008 overloads.

The only thing left unresolved, is recursive expression coverage in those cases like:
```
(a and b) xor (c and d).
```
Where I don't really know how to query the size of the array from the ```(a and b)``` sub-result.

**Comment by nickg** [comment]
> Where I don't really know how to query the size of the array from the (a and b) sub-result.

You can use `lower_array_len()` but it won't necessarily be a compile-time constant. 

There's an awful lot of complexity here to handle the individual bits of the arrays: is this feature really useful enough to justify that? I can understand for scalar std_logic values it's useful to track coverage as with boolean, but is there a real world use case for tracking the individual bits of a vector like this?

**Comment by Blebowski** [comment]
I agree that the feature is complex. OTOH, I think it makes sense to implement it.
I can think of concrete use-case in DSP code (unfortunately, in proprietary code-base).
If it is too much "cover tags" spam, we can not collect it by default, and enable it
only upon demand (similar as with the `ignore`/`include-mems` discussion).

A good use-case for this feature is showing un-reachable / dead code (e.g. when two vectors
are `and`-ed together, but some of the bits of one of the operands are never turned
to 1, thus resulting bit never turns to 1. In such case, it is good to `and` only the "reachable"
parts of the vector. Synthesis may throw out unreachable stuff, but AFAIK only due
to combinatorial constant propagation / boundary optimization. If one of the operands
is a register which just "never gets to the given values in the application scenario",
this may show places for optimization.

As for the nested expressions, I would skip them in this case completely. Since Array length
will not be a compile time constant, I don't really know how to get number of cover tags
to emit.

In the end, its of course your call :)

**Comment by nickg** [comment]
Is this feature implemented by any commercial tool? I checked the ModelSim user guide but in the Condition and Expression Coverage section it says:

> Expressions whose result is greater than one bit wide are not counted for coverage; they are silently ignored.

**Comment by Blebowski** [comment]
I have evaluated the behavior with Synopsys VCS, and it also ignores logic ops between vectors.

As its the case within Modelsim and VCS, I will stick with that and remove the feature.

**Comment by Blebowski** [comment]
OK, I have removed the vectors, MR is now ready for a review.

**Review by nickg — APPROVED** [review]

**Pull request merged**